### DISCUSSION TITLE
convenient_test_manager_dart runs forever when an assertion error occurs

### DISCUSSION THREAD

**Discussion by lehmannk** [fzyzcjy/flutter_convenient_test#350] [Q&A] [answered]
I have a very simple Test which should produce an assertion error because "DOES NOT EXIST" does not exist in the widget tree:
```convenientTestMain(MyConvenientTestSlot(), () {
    group('simple test group', () {
      tTestWidgets('simple test', (t) async {
        await t.get(find.text('DOES NOT EXIST')).should(findsOneWidget);
      });
    });
  });
```

Now, while the Manager UI (convenient_test_manager) properly detects the missing widget and prints the assertion failure the CLI version for CI `convenient_test_manager_dart` runs forever and never terminates. Here is the log:
```
2023-10-30T14:28:48.447752Z|debug|ReportHandlerService|handleReportLogEntry called
2023-10-30T14:28:48.572984Z|debug|ReportHandlerService|Snapshot
2023-10-30T14:28:48.775753Z|debug|ReportHandlerService|StateChange: testName=simple test group simple test state=status: complete
result: error
2023-10-30T14:28:48.791420Z|debug|ReportHandlerService|Error: 'package:flutter_test/src/binding.dart': Failed assertion: line 950 pos 14: '_pendingExceptionDetails != null': A test overrode FlutterError.onError but either failed to return it to its original state, or had unexpected additional errors that it could not handle. Typically, this is caused by using expect() before restoring FlutterError.onError. stack=dart:core-patch/errors_patch.dart 51:61       _AssertionError._doThrowNew
dart:core-patch/errors_patch.dart 40:5        _AssertionError._throwNew
package:flutter_test/src/binding.dart 950:14  TestWidgetsFlutterBinding._runTest.handleUncaughtError
package:flutter_test/src/binding.dart 955:9   TestWidgetsFlutterBinding._runTest.<fn>
dart:async/zone.dart 1081:14                  _Zone._processUncaughtError
dart:async/zone.dart 1285:5                   _CustomZone.handleUncaughtError
dart:async/future_impl.dart 790:16            Future._propagateToListeners
dart:async/future_impl.dart 649:5             Future._completeError
dart:async/zone.dart 1423:47                  _rootRunBinary
dart:async/zone.dart 1315:19                  _CustomZone.runBinary
dart:async/future_impl.dart 171:22            _FutureListener.handleError
dart:async/future_impl.dart 852:47            Future._propagateToListeners.handleError
dart:async/future_impl.dart 873:13            Future._propagateToListeners
dart:async/future_impl.dart 649:5             Future._completeError
dart:async/zone.dart 1407:47                  _rootRunUnary
dart:async/zone.dart 1308:19                  _CustomZone.runUnary
dart:async/zone.dart 1399:13                  _rootRun
dart:async/zone.dart 1301:19                  _CustomZone.run
dart:async/zone.dart 1209:7                   _CustomZone.runGuarded
dart:async/zone.dart 1249:23                  _CustomZone.bindCallbackGuarded.<fn>
dart:async/schedule_microtask.dart 40:21      _microtaskLoop
dart:async/schedule_microtask.dart 49:5       _startMicrotaskLoop
2023-10-30T14:28:48.791420Z|debug|ReportHandlerService|handleReportLogEntry called
🚀 [STATUS] [0:16] [1x completeFailureOrError] [Test=simple test group simple test]
🚀 [STATUS] [0:18] [1x completeFailureOrError] [Test=simple test group simple test]
2023-10-30T14:28:53.244813Z|info|main|monitorWorkerAvailable check
🚀 [STATUS] [0:20] [1x completeFailureOrError] [Test=simple test group simple test]
🚀 [STATUS] [0:22] [1x completeFailureOrError] [Test=simple test group simple test]
🚀 [STATUS] [0:24] [1x completeFailureOrError] [Test=simple test group simple test]
2023-10-30T14:28:58.260562Z|info|main|monitorWorkerAvailable check

... [running forever from here]
```

What am I doing wrong? AND: observing the logs in a CI build, what is the best trigger to detect a successful testrun or a failed testrun? The manager logs are full of debug messages etc. It is very difficult to actually identify what went wrong with regards to test assertions. How can I turn debug logging off for example?

**Comment by fzyzcjy** [comment]
That looks like a bug! If you only have one test, `🚀 [STATUS] [0:24] [1x completeFailureOrError] [Test=simple test group simple test]` means that, it successfully knows there is one failed (errored) test. Looks like the run status https://github.com/fzyzcjy/flutter_convenient_test/blob/c24191d1298b1f969b2053b8b98891d58161cf15/packages/convenient_test_manager_dart/bin/convenient_test_manager_dart.dart#L102 is somehow not updated.

> observing the logs in a CI build, what is the best trigger to detect a successful testrun or a failed testrun?

By nonzero exit code, it is a failed run

> The manager logs are full of debug messages etc. It is very difficult to actually identify what went wrong with regards to test assertions. How can I turn debug logging off for example?

The option to control it is in https://github.com/fzyzcjy/flutter_convenient_test/blob/c24191d1298b1f969b2053b8b98891d58161cf15/packages/convenient_test_common_dart/lib/src/common_dart/front_log.dart#L16C10-L16C10. Feel free to PR to make it configurable from outside! (Or hack the code and modify it directly)

**Reply by lehmannk** [reply]
Ok, I added another passing test. So I have 2 now - one passing and one failing. However, whereever I put my two tests (either 2 separate `groups`, 2 separate `tTestWidgets` or both `await t.get(...).should(...)` inside a single `tTestWidgets` closure the `convenient_test_manager_dart` command never quits and loops forever as shown above. This is quite critical for CI completion. Can you fix this soon?

**Reply by fzyzcjy** [reply]
I see. This looks not super hard to fix (just follow the data flow and see why that variables is not changed). But unfortunately, I am busy with things related to research and projects, so do not have time to fix it :/ Feel free to PR and I am willing to review and merge it :)

**Reply by lehmannk** [reply]
I tried to understand what is going on in this particular code block but I could not get it. I will review this with a fresh mind tomorrow I think. Besides, I noticed that the testmanager_dart produces 2 result folders (with result.bin each) whenever I run a single test with
```
flutter run .\test\integration_test\mini_test.dart -d windows --host-vmservice-port 9753 --disable-service-auth-codes --dart-define CONVENIENT_TEST_APP_CODE_DIR=lib\app
```
The first result folder always contains an empty report:
<img width="565" alt="image" src="https://github.com/fzyzcjy/flutter_convenient_test/assets/36925534/c07728cb-8cd7-4927-947b-29bdcf0d14b5">
The second one is the proper one.

May this have to do with the issue here (testmanager not tearing down tests properly)?
Because: the testmanager UI version does tear down all tests properly (also with failing tests) and the UI version also only produces a single report folder as apposed to the manager_dart cli version.

**Reply by fzyzcjy** [reply]
That looks possible!

Btw, just to ensure we are checking on a minimal reproducible sample, ensure we do not have fancy `tearDownAll` etc in the test script

**Reply by lehmannk** [reply]
The sample to reproduce is very simple indeed, nothing special at all:
```
void main() {
  convenientTestMain(MyConvenientTestSlot(), () {
    group('simple test group', () {
      tTestWidgets('simple test 1', (t) async {
        await t.get(find.text('blahh')).should(findsOneWidget);
      });
    });

  });
}
```
It is when the widget is not found when the manager_dart does not tear down anymore. Also, with this sample you always get 2 report folders with manager_dart instead of one.

**Reply by lehmannk** [reply]
got a little forward with my analysis. When a widget is not found (simple sample above) then an exception is thrown. This exception then stops the testmanager_dart from proceeding properly and it runs forever (not tearing down). When the exception occurs the following log entry is produced: 
```
2023-10-30T14:28:48.791420Z|debug|ReportHandlerService|Error: 'package:flutter_test/src/binding.dart': Failed assertion: line 950 pos 14: '_pendingExceptionDetails != null': A test overrode FlutterError.onError but either failed to return it to its original state, or had unexpected additional errors that it could not handle. Typically, this is caused by using expect() before restoring FlutterError.onError. stack=dart:core-patch/errors_patch.dart 51:61       _AssertionError._doThrowNew
```
In fact you can easily reproduce this behavior by simply throwing any exception, e.g.
```
  convenientTestMain(MyConvenientTestSlot(), () {
    group('simple test group', () {
      tTestWidgets('simple test 1', (t) async {
        throw Exception();
      });
    });
```
There seems to be an issue when overwriting `FlutterError.onError`, see e.g. here: 
 https://github.com/flutter/flutter/issues/34499#issuecomment-1381789793. Overwriting happens in convenient_test_dev in a single place here: https://github.com/fzyzcjy/flutter_convenient_test/blob/c24191d1298b1f969b2053b8b98891d58161cf15/packages/convenient_test_dev/lib/src/utils/errors.dart#L14-L20 
Does this cause the issue? Because the onError is not set back to its original state? The above mentioned exception ("A test overrode FlutterError.onError but...") appears also in the UI testmanager.

BTW: I tried to run testmanager_dart in isolation mode with arg `--isolation-mode` but the behavior stays the same.

**Reply by fzyzcjy** [reply]
Good job!

> The above mentioned exception ("A test overrode FlutterError.onError but...") appears also in the UI testmanager.

One strategy may be, since we know UI works while headless mode is buggy, the bug may be in the places where the two cases disagree with each other. Thus, the log may not be a problem if both cases have it.
For example, personally speaking, I may add a log of `print`s, and see how the two cases differ (e.g. one case prints out sth while the other case does not).

**Reply by lehmannk** [reply]
In fact, I think both UI and headless have the same problem - both get affected by "uncaught" exceptions and print it -> UI on the screen, headless is writing everything to reportfile properly BUT has the problem of never terminating. I think as the UI is not supposed to terminate after test end we do not see the same issue there. My guess really is that the root cause for all this sits in the above mentioned exception. 

Side note: When throwing an exception right at the beginning of a test (inside `tTestWidgets` then the whole test terminates at this point no matter how many other following tests, `tTestWidgets` or even  there are `groups` there are. Activating ISOLATION_MODE does not affect this behavior so this wouldn't help either - which I hoped would. (In which cases does isolation work then if not in uncaught exceptions?)

So I guess fixing the "A test overrode FlutterError.onError but..." error would help in all cases.

**Reply by lehmannk** [reply]
I think I found the solution, it is a general problem with integration_test and overriding `FlutterError.onError` the solution is given in https://github.com/flutter/flutter/issues/34499#issuecomment-1381789793. I have implemented the fix in `convenient_test_dev` locally and now exceptions are properly handled. Also, the CLI manager terminates properly in case of errors. I will prepare a PR for this tomorrow ;-)

**Reply by fzyzcjy** [reply]
Great job!

**Reply by lehmannk** [reply]
I cannot push my bugfix branch `bugfix/prevent_onError_override_exceptions`, do you need to set any permissions for me?
```
Push failed remote: Permission to fzyzcjy/flutter_convenient_test.git denied to lehmannk. unable to access 'https://github.com/fzyzcjy/flutter_convenient_test.git/': The requested URL returned error: 403
```

**Reply by fzyzcjy** [reply]
IMHO a standard workflow in GitHub involves things like "pull request" and "fork", https://github.com/features/code-review

**Reply by lehmannk** [reply]
oh sorry, I wasn't aware I need to fork first. Here is my PR: https://github.com/fzyzcjy/flutter_convenient_test/pull/351

**Comment by lehmannk** [comment, marked as answer]
Ok, for everyone interested. Here is the solution to avoid the `onError` problem and endless headless runs. Simply backup & restore the `onError`handler in your `ConvenientTestSlot` override:

```
 @override
  Future<void> appMain(AppMainExecuteMode mode) async {
    // onError overrides need to be restored in order to let the headless test manager terminate properly in case of exceptions
    // see https://github.com/fzyzcjy/flutter_convenient_test/pull/351
    FlutterExceptionHandler? originalOnError = FlutterError.onError;
    await app.main();
    FlutterError.onError = originalOnError;
  }
```

**Reply by fzyzcjy** [reply]
Good job! 🎉
